cryptolib/signer.go

```diff
@@ -20,37 +20,51 @@ import "errors"
 
 // Signer contains methods to create a signed Attestation.
 type Signer interface {
-	CreateAttestation(privateKey, payload []byte, publicKeyID string, alg SigningAlgorithm, keyMode KeyMode) (*Attestation, error)
+	// CreateAttestation creates an Attestation whose signature is generated by
+	// signing the given payload with the private key. For PGP and PKIX, `payload`
+	// should be the raw payload data. For JWT, `payload` should be a serialized
+	// but unsigned token.
+	CreateAttestation(payload []byte, alg SignatureAlgorithm) (*Attestation, error)
 }
 
-type signer struct{}
+type signer struct {
+	PrivateKey  []byte
+	PublicKeyID string
+}
+
+// NewSigner creates a Signer interface. `privateKey` contains the key data for
+// the private key that signs attestations. `publicKeyID` is the ID of the
+// public key that can verify those signatures.
+func NewSigner(privateKey []byte, publicKeyID string) Signer {
+	return &signer{
+		PrivateKey:  privateKey,
+		PublicKeyID: publicKeyID,
+	}
+}
 
-// CreateAttestation creates an Attestation whose signature is generated by
-// signing the given payload with the private key. For PGP and PKIX, `payload`
-// should be the raw payload data. For JWT, `payload` should be a serialized
-// but unsigned token.
-func (s *signer) CreateAttestation(privateKey, payload []byte, publicKeyID string, alg SigningAlgorithm, keyMode KeyMode) (*Attestation, error) {
-	switch keyMode {
+// CreateAttestation creates a signed Attestation. See Signer for more details.
+func (s *signer) CreateAttestation(payload []byte, alg SignatureAlgorithm) (*Attestation, error) {
```

Comment:
Nit: How common is it to use multiple algorithms with a single key? If that is not common I think it makes sense to move the alg field into NewSigner.

Reply:
Done. To be clear, each key mode's signer is now implemented by its own type. We added the SignatureAlgorithm parameter to the PKIX and JWT constructors. We did not include it in the PGP constructor, as the algorithm can be inferred directly from the private key.